servoshell: Port Android / OHOS servoshell to use the `WebView` API

mrobinson opened the pull request:
---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes do not require tests because they do not have any functional changes.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
servoshell: Port Android / OHOS servoshell to use the WebView API

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mukilan reviewed (changes requested):
I tested this branch on Android and I discovered an issue - the servo.org homepage doesn't is not loaded when launching the app and `about:blank` is loaded instead.

I debugged the code and found the issue is not related to this PR, but introduced in #34966 where `pref.homepage` was replaced with `servoshell_preferences.url`. The latter represents the parsed url provided via CLI, which would be none.

This patch fixes the issue for me:
```diff
diff --git a/ports/servoshell/egl/android/simpleservo.rs b/ports/servoshell/egl/android/simpleservo.rs
index 195a929bce..9f65deffc4 100644
--- a/ports/servoshell/egl/android/simpleservo.rs
+++ b/ports/servoshell/egl/android/simpleservo.rs
@@ -71,10 +71,7 @@ pub fn init(
     };

     let embedder_url = init_opts.url.as_ref().and_then(|s| ServoUrl::parse(s).ok());
-    let pref_url = servoshell_preferences
-        .url
-        .as_ref()
-        .and_then(|s| ServoUrl::parse(s).ok());
+    let pref_url = ServoUrl::parse(&servoshell_preferences.homepage).ok();
     let blank_url = ServoUrl::parse("about:blank").ok();

     let url = embedder_url.or(pref_url).or(blank_url).unwrap()
```

mrobinson commented:
> I debugged the code and found the issue is not related to this PR, but introduced in #34966 where `pref.homepage` was replaced with `servoshell_preferences.url`. The latter represents the parsed url provided via CLI, which would be none.

Yep, I noticed this as well. I hope to create a PR to fix it today as it was unrelated.

mrobinson commented on ports/servoshell/egl/servo_glue.rs:
```diff
@@ -515,39 +530,41 @@ impl ServoGlue {
                     }
                 },
                 EmbedderMsg::AllowOpeningWebView(response_chan) => {
-                    // Note: would be a place to handle pop-ups config.
```
It's still relevant, but these kind of comments will be moved to rustdoc on the delegate methods rather than as inline comments in servoshell, which will likely never have a popup blocker.

mrobinson commented on ports/servoshell/egl/android.rs:
```diff
@@ -43,11 +43,14 @@ pub extern "C" fn android_main() {
+                (f)(servo);
+                Ok(())
+            },
             None => Err("Servo not available in this thread"),
```
I've done this now.

mrobinson commented on ports/servoshell/egl/servo_glue.rs:
```diff
@@ -171,250 +181,259 @@ impl ServoGlue {
+            Point2D::new(x, y),
+            TouchEventType::Up,
+        );
+        self.maybe_perform_updates()
```
I think we are going to remove the optional batching from the Servo API. It seems like this is something that the application should handle. The fact that it was in the old API probably pointed to some design issues in any case.

mrobinson commented on ports/servoshell/egl/android.rs:
```diff
@@ -152,7 +155,7 @@ pub extern "C" fn Java_org_servo_servoview_JNIServo_setBatchMode<'local>(
 ) {
     debug!("setBatchMode");
-    call(&mut env, |s| s.set_batch_mode(batch == JNI_TRUE));
+    call(&mut env, |s| s.set_batch_mode(batch == JNI_TRUE).unwrap());
```
Thanks for pointing this out. `setBatchMode` and `set_batch_mode` were completely unused, so I've removed them. I think they aren't necessary any longer.

mrobinson commented:
> > I debugged the code and found the issue is not related to this PR, but introduced in #34966 where `pref.homepage` was replaced with `servoshell_preferences.url`. The latter represents the parsed url provided via CLI, which would be none.
> 
> Yep, I noticed this as well. I hope to create a PR to fix it today as it was unrelated.

I've gone ahead and fixed this one here. In the end, I've moved URL processing into the shared code, which should make things simpler and make it easier to test this in the future.

mukilan commented:
> https://github.com/servo/servo/actions/runs/13026616898

Looks like OHOS build passed but not HarmonyOS. I'll trigger a new try run.

mukilan commented:
> ⚠️ Try run ([#13026969442](https://github.com/servo/servo/actions/runs/13026969442)) failed.

Android is failing due to missing import statement:
```
error[E0412]: cannot find type `EmbedderProxy` in this scope
   --> ports/servoshell/egl/servo_glue.rs:701:26
    |
701 |         _embedder_proxy: EmbedderProxy,
    |                          ^^^^^^^^^^^^^ not found in this scope
```


> ✨ Try run ([#13027248559](https://github.com/servo/servo/actions/runs/13027248559)) succeeded.

OHOS is passing and HarmonyOS is still failing, but this should not block the merge. @jschwe It looks like `hvigor` is failing to launch java on the HOS runner.

```
2025-01-29T08:25:38.0138947Z > hvigor [91mERROR: Failed :servoshell:default@PackageHap... [39m
2025-01-29T08:25:38.0141859Z > hvigor [91mERROR: spawn java ENOENT[39m
2025-01-29T08:25:38.0158763Z > hvigor [91mERROR: BUILD FAILED in 3 s 228 ms [39m
```

mrobinson commented:
> > ⚠️ Try run ([#13026969442](https://github.com/servo/servo/actions/runs/13026969442)) failed.
> 
> Android is failing due to missing import statement:
> 
> ```
> error[E0412]: cannot find type `EmbedderProxy` in this scope
>    --> ports/servoshell/egl/servo_glue.rs:701:26
>     |
> 701 |         _embedder_proxy: EmbedderProxy,
>     |                          ^^^^^^^^^^^^^ not found in this scope
> ```

This should be fixed in the latest version of the branch.

jschwe commented:
> OHOS is passing and HarmonyOS is still failing, but this should not block the merge. @jschwe It looks like hvigor is failing to launch java on the HOS runner.

Thanks for the heads-up. That happened while I was refactoring https://github.com/servo/ci-runners/pull/22 and thought the `builder` image didn't need java.
